x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/TestCaseSupplier.java

```diff
@@ -1177,6 +1188,12 @@ public TestCase withWarning(String warning) {
             }
             return new TestCase(data, evaluatorToString, expectedType, matcher, newWarnings, expectedTypeError);
         }
+
+        public <T extends Throwable> TestCase withFoldingException(Class<T> clazz, String message) {
```

Comment:
Huh.  I'm not sure why we need this; or rather, I'm not sure why we didn't need this sooner if we need it at all. When are the folding warnings different from the evaluator warnings? is it possible we can merge down to a single warnings/exceptions function some how?

Reply:
> I'm not sure why we need this; or rather, I'm not sure why we didn't need this sooner if we need it at all.

It's been needed, IMO, we just didn't do it (for instance, have a look at 0-avoiding random number generation in `DivTests`, but no 0 devisor test.. which this PR also adds).
I've added this now because some functions (like `DateParseTests.java`) generate the same-ish message both in `Warning`s (when run through the execution engine) and folding failures (when the expression can be executed / folded before execution) and the folding tests cannot (and should not) be disabled.